src/Common/src/System/Net/Http/TlsCertificateExtensions.cs

```diff
@@ -0,0 +1,133 @@
+                            {
+                                hasClientCertExt = true;
+                                break;
+                            }
+                        }
+                    }
+                }
+
+                /// Check if the extension is a key usage extension.
+                /// No point going over it if we have already established that out cert has digital signature
+                if (!forDigitalSignature)
+                {
+                    X509KeyUsageExtension usageExt = extension as X509KeyUsageExtension;
+                    if (usageExt != null)
+                    {
+                        noUsageRestrictions = false;
+                        forDigitalSignature = (usageExt.KeyUsages & X509KeyUsageFlags.DigitalSignature) != 0;
+                    }
+                }
+
+                if(hasClientCertExt && forDigitalSignature)
+                {
+                    return true;
+                }
+            }
+            Debug.Assert(!hasClientCertExt || !forDigitalSignature, "if cert can be used as digital signature && it has ClientCertificateOid extension, it is a client cert");
+
+            return noExtendedRestrictions && noUsageRestrictions;
+        }
+
+        /// <summary>
+        ///   Returns a new collection containing valid client certificates from the given X509Certificate2Collection
+        /// </summary>
+        internal static bool TryFindClientCertificate(this X509Certificate2Collection certificates,
+                                                      ISet<string> allowedIssuers,
+                                                      out X509Certificate2 clientCertificate,
+                                                      out X509Chain clientCertChain)
+        {
+            clientCertificate = null;
+            clientCertChain = null;
+            if (certificates == null)
+            {
+                return false;
+            }
+
+            DateTime now = DateTime.Now;
+            foreach(X509Certificate2 cert in certificates)
+            {
+                if (cert.HasPrivateKey && now > cert.NotBefore && now < cert.NotAfter)
+                {
+                    if (IsClientCertificate(cert))
+                    {
+                        X509Chain chain = new X509Chain();
+                        chain.ChainPolicy.VerificationFlags = X509VerificationFlags.NoFlag;
+                        chain.ChainPolicy.RevocationFlag = X509RevocationFlag.ExcludeRoot;
+                        chain.ChainPolicy.ApplicationPolicy.Add(new Oid(ClientCertificateOid));
+                        if (!chain.Build(cert))
+                        {
+                            // TODO: uncomment the 'continue;'
+
+                            // Ideally, we would like to set VerificationFlags = AllFlags
+                            // But we are getting "work in progress" exception here when verificationFlags is anything but NoFlag
```

Comment:
Okay, better question.  Why do you want to ignore all of those potential errors?  What do you actually want to check?

https://msdn.microsoft.com/en-us/library/system.security.cryptography.x509certificates.x509verificationflags%28v=vs.110%29.aspx

Note that "AllFlags" is "include every Ignore", not "test everything".